src/binary/merkle_tree.rs

```diff
@@ -135,20 +141,123 @@ where
+    ///
+    /// ```
+    ///
+    /// No additional intermediate nodes or leaves are required to calculate
+    /// the root position.
+    ///
+    /// The positions of the MMR peaks can be deterministically calculated as a
+    /// function of `n + 1` where `n` is the number of leaves in the tree. By
+    /// appending an additional leaf node to the tree, we generate a new tree
+    /// structure with additional internal nodes (N.B.: this may also change the
+    /// root position if the tree is already balanced).
+    ///
+    /// In our example, we add an additional leaf at leaf index `7` (in-order
+    /// index `14`).
+    ///
+    /// ```text
+    ///               07
+    ///              /  \
+    ///             /    \
+    ///            /      \
+    ///           /        \
+    ///          /          \
+    ///         /            \
+    ///       03              11
+    ///      /  \            /  \
+    ///    ...  ...         /    \
+    ///                   09      13
+    ///                  /  \    /  \
+    ///                ...  ... 12  14
+    /// ```
+    ///
+    /// We observe that the path from the root position to our new leaf position
+    /// yields a set of side positions that includes our original peak
+    /// positions (see [Path Iterator](crate::common::path_iterator::PathIter)):
+    ///
+    /// | Path position | Side position |
+    /// |---------------|---------------|
+    /// |            07 |            07 |
+    /// |            11 |            03 |
+    /// |            13 |            09 |
+    /// |            14 |            12 |
+    ///
+    /// By excluding the root position `07`, we have established the set of
+    /// side positions `03`, `09`, and `12`, matching our set of MMR peaks.
+    ///
     fn build(&mut self) -> Result<(), MerkleTreeError<StorageError>> {
-        let keys = (0..self.leaves_count).map(|i| Position::from_leaf_index(i).in_order_index());
-        for key in keys {
+        // Define a new tree with a leaf count 1 greater than the current leaf
+        // count.
+        let leaves_count = self.leaves_count + 1;
+
+        // Get the rightmost leaf position of the new tree. The rightmost leaf
+        // position is always the position with leaf index N - 1, where N is
+        // the number of leaves.
+        let leaf_position = Position::from_leaf_index(leaves_count - 1);
+
+        // The root position of a tree will always have an in-order index equal
+        // to N - 1, where N is the leaves count rounded (or equal) to the next
+        // power of 2.
+        let root_index = leaves_count.next_power_of_two() - 1;
+        let root_position = Position::from_in_order_index(root_index);
+
+        let (_, peaks): (Vec<_>, Vec<_>) = root_position
+            .path(&leaf_position, leaves_count)
+            .iter()
+            .unzip();
+
+        let mut current_head = None;
+        let peaks = &peaks.as_slice()[1..]; // Omit the root.
```

Comment:
Is this index safe?

Reply:
Yes, this will be safe because the vector is guaranteed to contain at least the root position. We ask a node for its path to another node by calling a member function. The minimum path size of 1 occurs when both nodes in the query are actually the same node. In this case, that will be the root.

Reply:
Awesome, maybe just add a comment to that effect